### PR TITLE
zoekt-webserver: add JSON API

### LINKED ISSUE #410: RPC: support non-gob encodings?
Hi.  I'd like to slap an alternative web UI on zoekt (example motivations: instant search, like livegrep; syntax highlighting in the builtin file viewer; alternative homepage suitable to my particular deployment).  The APIs exposed in web (when passing `format=json`) don't seem sufficient to do this.  For example, there doesn't seem to be a way to list repositories, or to retrieve full file contents.

Both of those things are provided by rpc/.  But rpc only speaks gob.  It seems the only way to use rpc from not-go is to proxy it with go, translating into something that's not gob.  I could do that, but it feels like a waste.  Ideally there would be a way to serve another more portable encoding (JSON seems like the least controversial option).

I'm no go expert, but it doesn't seem trivial to serve JSON instead of gob, in the way that it is to serve JSON instead of HTML in web, because rpc directly uses `net/rpc` (or, your fork of it), which precludes alternative encodings.  Does that sound correct?

Thanks!

### PR DESCRIPTION
**Pull request by isker**
This adds a JSON API that directly exposes the Search and List functions
at /api/search and /api/list in zoekt-webserveer, gated behind the
existing -rpc flag.

Closes #410.

### CODE REVIEW

File: rpc/internal/srv/srv.go
Comment (isker) [inline rpc/internal/srv/srv.go]: `rpc` is not necessarily an appropriate package for all of this to live in.  Suggestions welcome.

File: rpc/internal/srv/srv.go
Comment (isker) [inline rpc/internal/srv/srv.go]: I have not downcased these (e.g. `Q` -> `q`) with JSON tags because it'd be inconsistent for these to be downcased but `zoekt.ListOptions` or `zoekt.SearchOptions` not to be (and so on, recursively).  I can of course downcase all of that if that's the right thing to do.

File: web/server.go
Comment (isker) [inline web/server.go]: Bikeshedding on URL paths welcome.

**Review by isker — COMMENTED** [review]
@keegancsmith if this looks generally acceptable to you, I can write actual test cases.  I have done simple tests locally on my zoekt checkout:

```
λ echo -n '{"Q": "foobar"}' | xh post localhost:6070/api/search | jq -r .Result.Files[0].LineMatches[0].Line | base64 -d
	in := "[a-zA-Z]fooBAR"⏎   
```

**Comment by isker** [comment]
Any thoughts on this PR?  I'd ping the "Seach core" team @keegancsmith assigned but it seems like it's private or something else prevents me from doing so 🌞.

**Comment by stefanhengl** [comment]
I am going to take a look. I had a brief look already but I won't be able to finish my review today.

File: rpc/internal/srv/srv.go
Comment (stefanhengl) [reply rpc/internal/srv/srv.go]: I agree, this should live someplace else.

**Review by stefanhengl — CHANGES_REQUESTED** [review]
Thank you for working on this!  I think the way to go is to extend `serveSearch` rather than creating a separate API. For example, we could change `serveSearchErr` to return `zoekt.SearchResult` instead of `ApiSearchResult` and, in `serveSearch`, switch based on the requested format. Does this make sense?

I imagine as part of this refactor you have to pull out a converter `zoekt.SearchResult -> ApiSearchResult` from `serveSearchErr` and call it in `serveSearch`.  

To maintain backward compatibility we would have to keep the existing "json" as is (we know there are users relying on this) and add another format, EG "json2", "json_new" (I obviously have no imagination when it comes to names :-)).

WDYT?

**Comment by isker** [comment]
@stefanhengl this is possible to do, I think, but there are some questions to be answered.  I didn't implement this in `serveSearchErr` to start with because it has a lot of strange behaviors compared to the gob API that I wanted to avoid.

For example, it takes an entirely different behavior if it thinks you're only querying for repos: https://github.com/sourcegraph/zoekt/blob/2478bf5713dc12a48b21f1cd487bb3ef7ee3f99b/web/server.go#L245-L264

There's no such behavior in the gob API.

It goes from there to constrain or modify the request in ways that seem tightly fitted toward satisfying the HTML interface of the webserver.  It really does feel like the existing JSON API was bolted onto this function after the fact.  I'm aiming for a clean break, which is why I have straightforwardly exposed the request and response API objects in this PR.

Additionally, you described how we might refactor to vary the response format, but what do we do about the request format?  Are you suggesting to not expose `zoekt.SearchOptions` in this API, instead just accepting the relatively limited inputs `serveSearchErr` does today?  Or should our new `format` value also fork request parsing behavior?

Consider that `serveSearchErr` serves GET requests with a query string; I'd want the new API to be served over POST with a JSON body, both for HTTP semantic reasons (this really is a remote procedure call that does not fit into the model of idempotency implied by GET), and, much more importantly, for practical implementation reasons (`zoekt.SearchOptions` is not flat, and so is not trivially parsable out of a query string).

**Comment by stefanhengl** [comment]
Ok. I suggest to create a separate package `json` at the root of the repository. This way you loose all your `JSON` prefixes and the code stays nice and contained. You can move all your code from rpc to there. 

As you have seen, we have two other API routes, `rpc` and `stream`, which serve the same content. I am not sure if `/api` is most intuitive route for the JSON API. Maybe @keegancsmith has a better idea?

**Comment by isker** [comment]
I've moved the JSON endpoint into its own package as suggested.  I also added a test mirroring that in rpc.

The only functional change is that I've relaxed the /list endpoint to allow a missing query, such that all repos will be returned.  I think it makes sense to want to list all repos this way, though the equivalent in /search remains nonsense.

File: json/json.go
Comment (keegancsmith) [inline json/json.go:88]: Can you call `searchArgs.Opts.SetDefaults()` before calling Search.

File: json/json.go
Comment (keegancsmith) [inline json/json.go:67]: minor: in the webserver handler it does some adjustments to SearchOptions based on the number of results it expects to find which is pretty neat. This may be useful behaviour to have in this handler if opts is not set: https://github.com/sourcegraph/zoekt/blob/d1964a317f841afecc71b7210fb6ef4c3de26733/web/server.go#L290-L316

File: web/server.go
Comment (keegancsmith) [inline web/server.go:177]: nit add a variable to `json` for `DefaultAPIPath` to match the style of the other endpoints.

**Review by keegancsmith — APPROVED** [review]
LGTM! Thanks so much. I'll let @stefanhengl  take a look over before merging in.

File: json/json.go
Comment (isker) [reply json/json.go:88]: This is already being done here, right?

https://github.com/sourcegraph/zoekt/blob/d1964a317f841afecc71b7210fb6ef4c3de26733/eval.go#L144-L147

Based on that, I think the call here is actually redundant.

https://github.com/sourcegraph/zoekt/blob/d1964a317f841afecc71b7210fb6ef4c3de26733/web/server.go#L288-L291

File: web/server.go
Comment (isker) [reply web/server.go:177]: Well, this line works only when passing exactly `/api/` to Handle and `/api` to StripPrefix.  I originally did what you said, but I thought string literals were much clearer in the end.

File: json/json.go
Comment (isker) [reply json/json.go:67]: I explicitly imitated what rpc does (or does not) do when implementing this endpoint, instead of imitating the webserver.  Am I correct in saying that rpc does not do this?  Does sourcegraph (as clients of rpc/stream) implement something similar, or?

Comment (keegancsmith) [reply json/json.go:67]: Yup RPC does not do this, so on sourcegraph's side we implement that behaviour. So doing it will improve the ergonomics for the client of this API.

**Review by stefanhengl — APPROVED** [review]
LGTM!

File: json/json.go
Comment (isker) [reply json/json.go:67]: I extracted a shared function that's used by both the new JSON endpoint and the webserver.  I put it in the new json package, but if you think there's a better place, let me know.

File: json/json.go
Comment (stefanhengl) [inline json/json.go]: nit: can we flip this around to reduce nesting?

Would this work?
```
	if opts.MaxDocDisplayCount == 0 || opts.ShardMaxMatchCount != 0 || opts.ShardMaxImportantMatch != 0 {
		return nil
	}

	maxResultDocs := opts.MaxDocDisplayCount
```

Comment (isker) [reply json/json.go]: Done.

**Comment by isker** [comment]
Thanks to you both!

**Pull request merged**
